### ISSUE TITLE
Means of accessing a tile from a tilemap collision and other related items

### ISSUE THREAD

**Issue by chucklepie** [godotengine/godot-proposals#1414]
**Describe the project you are working on:**
Platformer

**Describe the problem or limitation you are having in your project:**
Tilemap collisions provide no real meta data on a per tile basis, in addition when collisions occur all you get back is the entire Tilemap object (not the tile) and every tile has to share the same single collision mask/layer.

In order to get the tile something collides with, this is the best I've been able to do:

```
var tile_pos = collision.collider.world_to_map(position)
tile_pos -= collision.normal
var tile_id = collision.collider.get_cellv(tile_pos)
var one_way = collision.collider.tile_set.tile_get_shape_one_way (tile_id,0) 
```

It's not simple, takes a lot of thought and searching, and getting a collision for a tile should be a simple one liner in Godot. In addition, the 'shape_id' value '0' I used, I just guessed this was an index because I could find nothing on it.

As an aside, the documentation is little more than a few words for most properties in tilemap/tileset when trying to figure out collisions between tilemap and other objects, made more difficult by because collisions and colliders return generic object generic types, there is no auto-completion to help either.

I am in awe of the Godot developers, but the above should not be how you have to access a tile. It's painful and hard.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
1. Provide a means of accessing a tile from a collision or even replacing the 'collider' tilemap with the tile itself
2. Optional but really is required: provide a means for EACH tile to have its own collision mask/layer because a one for every single tile, which could be tens or hundreds in a tileset is a bit mad, leading on to
3. Optional, but really is required: a tile should have a means of providing some custom meta-data within it's own data

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
Somebody, better than me, who wrote the tilemap please make it simpler :)

**Is there a reason why this should be core and not an add-on in the asset library?:**
Tilemaps play a large part in 2D games and currently it's pretty hard to do a lot of things, making collisions better would be a start...

**Comment by TamerSoup625** [comment]
There is already a feature to get the tile you collide with, as stated in the docs (in TileMap API):
> When doing physics queries against the tilemap, the cell coordinates are encoded as metadata for each detected collision shape returned by methods such as Physics2DDirectSpaceState.intersect_shape(), Physics2DDirectBodyState.get_contact_collider_shape_metadata(), etc.

But, AFAIK, you can't get the metadata from a `KinematicCollision2D` and, rather than using high level nodes such as `RayCast2D`, you must use the space state's low level API.
At least being able to get the tile coordinates directly from a `KinematicCollision2D` (mabye with a general `metadata` property) would be a very great feature as it would not only be _easier_, but also _faster_.

**Comment by berarma** [comment]
The cell coordinates aren't that useful to know which tile caused the collision. See #90581.

**Comment by AThousandShips** [comment]
This is as mentioned above supported in 4.x, so closing as implemented in 4.0
